source/yats/include/yats/connection_helper.h

```diff
@@ -114,66 +137,157 @@ class connection_helper : public abstract_connection_helper
+    template <size_t Index = 0>
+    std::enable_if_t<Index == helper::output_count> add(size_t, void*)
     {
         throw std::runtime_error("Output Parameter locationId not found. It is not allowed to connect task from different pipelines.");
     }
 
-    template <size_t index = 0>
-    std::enable_if_t<(index < helper::input_count), void*> get(size_t location_id)
+    template <size_t Index = 0>
+    std::enable_if_t<(Index < helper::input_count), void*> get(size_t location_id)
     {
-        if (index == location_id)
+        if (Index == location_id)
         {
-            return &std::get<index>(m_callbacks);
+            return &std::get<Index>(m_callbacks);
         }
         else
         {
-            return get<index + 1>(location_id);
+            return get<Index + 1>(location_id);
         }
     }
 
-    template <size_t index = 0>
-    std::enable_if_t<index == helper::input_count, void*> get(size_t)
+    template <size_t Index = 0>
+    std::enable_if_t<Index == helper::input_count, void*> get(size_t)
     {
         throw std::runtime_error("Input parameter locationId not found. This implies an implementation error in yats.");
     }
 
+    /**
+    * Iterates through all inputs of the run function and stores the id of each input
+    * in m_input_ids. The index of the id corresponds to position of the input
+    * in parameters of the run function.
+    * <p>
+    * Note: This function should only be called once.
+    * </p>
+    */
+    void initialize_input_ids()
+    {
+        parse_input_ids(m_input_ids);
+    }
+
+    template <size_t Index = 0>
+    std::enable_if_t<Index < helper::input_count> parse_input_ids(std::vector<uint64_t>& ids)
+    {
+        ids.push_back(std::tuple_element_t<Index, input_tuple>::id);
+        parse_input_ids<Index + 1>(ids);
+    }
+
+    template <size_t Index = 0>
+    std::enable_if_t<Index == helper::input_count> parse_input_ids(std::vector<uint64_t>&)
+    {
+    }
+
+    /**
+     * Iterates through all outputs of the run function and stores the id of each output
+     * in m_output_ids. The index of the id corresponds to position of the output
+     * in the return value of the run function.
+     * <p>
+     * Note: This function should only be called once.
+     * </p>
+     */
+    void initialize_output_ids()
+    {
+        parse_output_ids(m_output_ids);
+    }
+
+    template <size_t Index = 0>
+    std::enable_if_t<Index < helper::output_count> parse_output_ids(std::vector<uint64_t>& ids)
+    {
+        ids.push_back(std::tuple_element_t<Index, output_tuple>::id);
+        parse_output_ids<Index + 1>(ids);
+    }
+
+    template <size_t Index = 0>
+    std::enable_if_t<Index == helper::output_count> parse_output_ids(std::vector<uint64_t>&)
+    {
+    }
+
     input_queue_ptr m_input;
     output_callbacks m_output;
     input_callbacks m_callbacks;
+    std::vector<uint64_t> m_input_ids;
+    std::vector<uint64_t> m_output_ids;
```

Comment:
I still do not like these members. They exist all the time but ar eonly needed when a specific function is called.

Reply:
Because the connection helper is used to get the inputs I find it legit to put the logic required to get the input_id to an input into the connection_helper class. After all it is just a helper class.

Reply:
It is bad style to keep variables that are unused in some use cases. Why you don't puts a function that returns the input/output vector and you call this on every connection_helper?

Reply:
Your wish is my command. I don't want to argue.